README.md
# SIMULACAO API

Este projeto é um simulador de empréstimos/financiamento, no qual o cliente envia para a API o `prazo` em meses que deseja pagar esse financiamento, e envia o `valorDesejado` desse financiamento.

A API irá retornar o valor das parcelas na tabela PRICE e na tabela SAC, e antes de enviar a resposta ao cliente existem diversas validações na API, mantendo assim a qualidade da resposta.

O projeto contêm logs, testes de integração, testes unitários e um documentação em Swagger da API.

A API foi construida em python usando o framework Flask e Flask-RestX. Todo o projeto está modularizado em blueprints permitindo inserir outras rotas e até outras APIs no mesmo projeto.

As configurações encontram-se no arquivo .env, é utilizado uma conexão com o SQLServer para buscar o produto especifico ao emprestimo desejado pelo cliente, e cada resposta da API é enviada também ao EventHub para analise estatísticas de outros departamentos.

## Pré-requisitos

Certifique-se de ter o [Python 3](https://www.python.org/downloads/)  instalado na sua máquina. 

Instale o driver do SQLServer versão 17 [`download`](https://go.microsoft.com/fwlink/?linkid=2223304) ele se encontra no seguinte [endereço](https://learn.microsoft.com/pt-br/sql/connect/odbc/download-odbc-driver-for-sql-server?view=sql-server-ver16#download-for-windows) juntamente com outras versões.

Após fazer o download do arquivo `msodbcsql.msi` basta executa-lo para instalar o drive.

Consulte o site da [Microsoft](https://learn.microsoft.com/pt-br/sql/connect/odbc/windows/system-requirements-installation-and-driver-files?view=sql-server-ver16#installing-microsoft-odbc-driver-for-sql-server) para mais informações.

## Instalação
Dentro da pasta raiz do projeto:
1. Crie o ambiente virtual:
```bash
python -m venv meu_venv
``` 
Irá aparecer no inicio da linha de comando a seguinte informação `(meu_venv)`

2.  Instale todas as dependências do projeto no ambiente virtual:
```bash
pip install -r requirements.txt
```
3.  Configure o ambiente virtual para utilizar o Flask executando o seguinte comando:
```bash
set FLASK_APP=main
```
4. Execute os testes para verificar que esta tudo funcionando com o comando:
```bash
flask test
```
Aparecendo `OK` no final significa que esta tudo funcionando.

5. Inicialize a aplicação com o comando:
```bash
flask run
```
6. No vscode (Visual Studio Code) instale a extençao `REST Client` para poder fazer testes no arquivo `test_api.http` diretamente no vscode.
  
7. No localhost o projeto funciona na porta 5000 `http://127.0.0.1:5000`
  
8. A documentação do projeto esta no formato `swagger` e para verificar e interagir com a API, com a aplicação rodando entre no seguinte endereço `http://127.0.0.1:5000/swagger`

CHANGELOG.md
# Changelog

Projeto "SIMULACAO API"

## [Unreleased]

## [1.0.0] - 2023-06-05

### Added

- Tabela Price
- Tabela SAC
- Documentação SWAGGER
- Modularização da aplicação
- Conexão com SQLServer
- Conexão com EventHub
- Geração de logs que encontram-se no arquivo app.log
- Testes automatizados
- Endpoint para realizar a simulação de financimento

tests/test_api.py
import unittest
import json
from flask_testing import TestCase
from main import app
from api.models.models import db, Produto
from api.config.config import configuration
from api.controllers.eventhub import enviar_json_ao_eventhub


class ApiTest(TestCase):
    def create_app(self):
        app.config.from_object(configuration["testing"])
        return app

    def setUp(self):
        try:
            db.create_all()
        except Exception as e:
            print("Erro: A conexão com o banco de dados não está aberta.", type(e).__name__)

    def test_database_connection(self):
        try:
            self.assertTrue(db.session.connection().closed == 0)
        except Exception as e:
            self.fail("Erro: A conexão com o banco de dados não está aberta.", type(e).__name__)

    def test_simulacao_endpoint(self):
        data = {"valorDesejado": 900.00, "prazo": 5}
        response = self.client.post("/simulacao/produto", json=data)
        self.assertEqual(response.status_code, 200)

    def test_simulacao_endpoint_method_get(self):
        response = self.client.get("/simulacao/produto")
        self.assertEqual(response.status_code, 405)

    def nao_retorna_produto(self, datas):
        contador = 0
        for data in datas:
            response = self.client.post("/simulacao/produto", json=data)
            response_data = json.loads(response.data.decode("utf-8"))

            if response_data["error"]["msg"] == "Produto nao encontrado":
                contador += 1

        return contador

    def retorna_produto(self, datas):
        resultado = []
        for data in datas:
            response = self.client.post("/simulacao/produto", json=data)
            response_data = json.loads(response.data.decode("utf-8"))
            produto = response_data["codigoProduto"]
            resultado.append({str(produto): data})

        return resultado

    def retorna_produto_schema_validar(self, datas):
        contador = 0
        for data in datas:
            response = self.client.post("/simulacao/produto", json=data)
            response_data = json.loads(response.data.decode("utf-8"))
            if list(response_data.keys()) == ['errors', 'message']:
                contador += 1

        return contador

    def test_estrutura_da_request(self):
        datas = [
            {"praz": 1, "valorDesejado": 200},
            {"prazo": "0r", "valorDesejado": 200}
        ]

        resultado = self.retorna_produto_schema_validar(datas)
        self.assertEqual(2, resultado)

    def test_simulacao_query(self):
        datas = [
            {"prazo": 1, "valorDesejado": 200},
            {"prazo": 1, "valorDesejado": 10000},
            {"prazo": 24, "valorDesejado": 200},
            {"prazo": 24, "valorDesejado": 10000},
            {"prazo": 25, "valorDesejado": 10000.01},
            {"prazo": 25, "valorDesejado": 100000.00},
            {"prazo": 48, "valorDesejado": 10000.01},
            {"prazo": 48, "valorDesejado": 100000.00},
            {"prazo": 49, "valorDesejado": 100000.01},
            {"prazo": 49, "valorDesejado": 1000000.00},
            {"prazo": 96, "valorDesejado": 100000.01},
            {"prazo": 96, "valorDesejado": 1000000.00},
            {"prazo": 97, "valorDesejado": 1000000.01},
            {"prazo": 97, "valorDesejado": 2000000.01},
            {"prazo": 100, "valorDesejado": 1000000.01},
            {"prazo": 100, "valorDesejado": 2000000.01}
        ]

        resultado = self.retorna_produto(datas)

        self.assertIn({"1": {"prazo": 1, "valorDesejado": 200}}, resultado)
        self.assertIn({"1": {"prazo": 1, "valorDesejado": 10000}}, resultado)
        self.assertIn({"1": {"prazo": 24, "valorDesejado": 200}}, resultado)
        self.assertIn({"1": {"prazo": 24, "valorDesejado": 10000}}, resultado)
        self.assertIn({"2": {"prazo": 25, "valorDesejado": 10000.01}}, resultado)
        self.assertIn({"2": {"prazo": 25, "valorDesejado": 100000.00}}, resultado)
        self.assertIn({"2": {"prazo": 48, "valorDesejado": 10000.01}}, resultado)
        self.assertIn({"2": {"prazo": 48, "valorDesejado": 100000.00}}, resultado)
        self.assertIn({"3": {"prazo": 49, "valorDesejado": 100000.01}}, resultado)
        self.assertIn({"3": {"prazo": 49, "valorDesejado": 1000000.00}}, resultado)
        self.assertIn({"3": {"prazo": 96, "valorDesejado": 100000.01}}, resultado)
        self.assertIn({"3": {"prazo": 96, "valorDesejado": 1000000.00}}, resultado)
        self.assertIn({"4": {"prazo": 97, "valorDesejado": 1000000.01}}, resultado)
        self.assertIn({"4": {"prazo": 97, "valorDesejado": 2000000.01}}, resultado)
        self.assertIn({"4": {"prazo": 100, "valorDesejado": 1000000.01}}, resultado)
        self.assertIn({"4": {"prazo": 100, "valorDesejado": 2000000.01}}, resultado)

    def test_produto_nao_encontrado(self):
        datas = [
            {"prazo": 25, "valorDesejado": 200},
            {"prazo": 1, "valorDesejado": 10000.01},
            {"prazo": 97, "valorDesejado": 10000.00},
            {"prazo": 49, "valorDesejado": 100000.00},
            {"prazo": -1, "valorDesejado": -200},
            {"prazo": -1, "valorDesejado": 200},
            {"prazo": 0, "valorDesejado": 200}
        ]

        resultado = self.nao_retorna_produto(datas)

        self.assertEqual(resultado, 7)

    def test_produto_columns(self):
        columns = Produto.__table__.columns.keys()

        self.assertIn("co_produto", columns)
        self.assertIn("no_produto", columns)
        self.assertIn("pc_taxa_juros", columns)
        self.assertIn("nu_minimo_meses", columns)
        self.assertIn("nu_maximo_meses", columns)
        self.assertIn("vr_minimo", columns)
        self.assertIn("vr_maximo", columns)

        self.assertEqual(str(Produto().__table__.c.co_produto.type), "INTEGER")
        self.assertEqual(str(Produto().__table__.c.no_produto.type), "VARCHAR(200)")
        self.assertEqual(
            str(Produto().__table__.c.pc_taxa_juros.type), "NUMERIC(10, 9)"
        )
        self.assertEqual(str(Produto().__table__.c.nu_minimo_meses.type), "SMALLINT")
        self.assertEqual(str(Produto().__table__.c.nu_maximo_meses.type), "SMALLINT")
        self.assertEqual(str(Produto().__table__.c.vr_minimo.type), "NUMERIC(18, 2)")
        self.assertEqual(str(Produto().__table__.c.vr_maximo.type), "NUMERIC(18, 2)")

        self.assertTrue(len(columns) == 7)

    def test_eventhub_connection(self):
        eventhub = enviar_json_ao_eventhub({"id": 1, "msg": "flask_test"})
        self.assertEqual(eventhub, "ok")


if __name__ == "__main__":
    unittest.main()


api/__init__.py
from flask import Flask, jsonify
from api.config.config import configuration
from api.routes.simulacao import simulacao, simulacao_ns
from .models.models import db
from flask_restx import Api


def create_app(config_name):
    app = Flask(__name__)
    app.config.from_object(configuration[config_name])
    app.config['SWAGGER_MASK_SWAGGER'] = False
    db.init_app(app)

    api = Api(
        simulacao,
        title='API Simulador',
        version='1.0',
        description='Documentacao da API de simulacao de emprestimo',
        doc='/swagger'
    )

    api.add_namespace(simulacao_ns)

    app.register_blueprint(simulacao)

    @app.errorhandler(404)
    def not_found_error(error):
        response = {"error": "Pagina nao encontrada", "status": 404}
        return jsonify(response), 404

    @app.errorhandler(405)
    def method_not_allowed_error(error):
        response = {"error": "Metodo nao permitido", "status": 405}
        return jsonify(response), 405

    return app


api/config/config.py
import os
from pathlib import Path
from dotenv import load_dotenv


env_path = Path(__file__).resolve().parent.parent.parent / ".env"
load_dotenv(env_path)


class Config(object):
    SECRET_KEY = os.environ.get("SECRET_KEY")
    DATABASE_URL = os.environ.get("DATABASE_URL")
    DB = os.environ.get("DB")
    PORT = os.environ.get("PORT")
    LOGIN = os.environ.get("LOGIN")
    PASSWORD = os.environ.get("PASSWORD")
    EVENTHUB_NAME = os.environ.get("EVENTHUB_NAME")
    EVENTHUB_ENDPOINT = os.environ.get("EVENTHUB_ENDPOINT")
    SWAGGER_MASK_SWAGGER = os.environ.get("SWAGGER_MASK_SWAGGER")


class DevelopmentConfig(Config):
    SQLALCHEMY_DATABASE_URI = (
        "mssql+pyodbc://{}:{}@{}:{}/{}?driver=ODBC+Driver+17+for+SQL+Server".format(
            Config.LOGIN, Config.PASSWORD, Config.DATABASE_URL, Config.PORT, Config.DB
        )
    )
    DEBUG = True


class TestingConfig(Config):
    SQLALCHEMY_DATABASE_URI = (
        "mssql+pyodbc://{}:{}@{}:{}/{}?driver=ODBC+Driver+17+for+SQL+Server".format(
            Config.LOGIN, Config.PASSWORD, Config.DATABASE_URL, Config.PORT, Config.DB
        )
    )
    TESTING = True
    WTF_CSRF_ENABLED = False
    PRESERVE_CONTEXT_ON_EXCEPTION = False


configuration = {
    "base": Config,
    "development": DevelopmentConfig,
    "testing": TestingConfig,
}


api/controllers/financiamento.py
def tabela_price(vp, n, i):
    if n == 0:
        return []

    vp_f = float(vp)
    n_f = float(n)
    i_f = float(i)

    z = (1.0 + i_f) ** n_f
    pmt = vp_f / ((z - 1.0) / (z * i_f))

    parcelas = []

    for periodo in range(n):
        juros = vp_f * i_f
        amortizacao = pmt - juros
        vp_f = vp_f - amortizacao

        parcela = {
            "numero": periodo + 1,
            "valorAmortizacao": float(round(amortizacao, 2)),
            "valorJuros": float(round(juros, 2)),
            "valorPrestacao": float(round(pmt, 2)),
        }

        parcelas.append(parcela)

    return parcelas


def tabela_sac(vp, n, i):
    if n == 0:
        return []

    vp_f = float(vp)
    n_f = float(n)
    i_f = float(i)

    amortizacao = vp_f / n_f

    parcelas = []

    for periodo in range(n):
        juros = vp_f * i_f
        pmt = amortizacao + juros
        vp_f = vp_f - amortizacao

        parcela = {
            "numero": periodo + 1,
            "valorAmortizacao": float(round(amortizacao, 2)),
            "valorJuros": float(round(juros, 2)),
            "valorPrestacao": float(round(pmt, 2)),
        }

        parcelas.append(parcela)

    return parcelas


api/routes/simulacao.py
from flask import Blueprint
from sqlalchemy.sql.expression import func
from ..models.models import Produto
from ..controllers.financiamento import tabela_price, tabela_sac
from ..controllers.eventhub import enviar_json_ao_eventhub
from flask import current_app

from log.log_config import logger
from flask_restx import Namespace, Resource, fields, reqparse, abort


# Construcao do Namespace Simulacao
simulacao_ns = Namespace('simulacao', 'Simulador API endpoints')

# Criando o modelo de request da api para o namespace Simulacao
simulacao_request_model = simulacao_ns.model('SimulacaoRequest', {
    'prazo': fields.Integer(
        required=True,
        description='Prazo do emprestimo'
    ),
    'valorDesejado': fields.Float(
        required=True,
        description='Valor desejado no emprestimo'
    )
})

# Criando o modelo de resposta, referente a Parcela
parcela_model = simulacao_ns.model('Parcela', {
    'numero': fields.Integer(
        readonly=True,
        description='Número da parcela'
    ),
    'valorAmortizacao': fields.Float(
        readonly=True,
        description='Valor de amortização da parcela'
    ),
    'valorJuros': fields.Float(
        readonly=True,
        description='Valor de juros da parcela'
    ),
    'valorPrestacao': fields.Float(
        readonly=True,
        description='Valor total da prestação da parcela'
    )
})

# Criando o modelo de resposta, referente ao Tipo, que engloba o modelo Parcela
tipo_modelo = simulacao_ns.model('Tipo', {
    'tipo': fields.String(
        readonly=True,
        description='Tipo da tabela de financiamento'
    ),
    'parcelas': fields.List(
        fields.Nested(parcela_model),
        readonly=True,
        description='Lista de parcelas'
    )
})

# Criando o modelo final da resposta, que engloba o modelo Tipo
simulacao_response_model = simulacao_ns.model('SimulacaoResponse', {
    'codigoProduto': fields.Integer(
        readonly=True,
        description='Código do produto'
    ),
    'descricaoProduto': fields.String(
        readonly=True,
        description='Descrição do produto'
    ),
    'taxaJuros': fields.Float(
        readonly=True,
        description='Taxa de juros'
    ),
    'resultadoSimulacao': fields.List(
        fields.Nested(tipo_modelo),
        readonly=True,
        description='Lista de tipos de financiamento'
    )
})

# Criando o modelo de resposta para o erro
erro_modelo = simulacao_ns.model('Erro', {
    'msg': fields.String(
        readonly=True,
        description='Bad Request'
    )
})

simulacao = Blueprint("main", __name__)

simulacao_request_parser = reqparse.RequestParser()
simulacao_request_parser.add_argument('prazo', type=int, required=True, help='Prazo do emprestimo')
simulacao_request_parser.add_argument('valorDesejado', type=float, required=True, help='Valor desejado no emprestimo')


@simulacao_ns.route("/produto", methods=["POST"])
class Simulacao(Resource):

    @simulacao_ns.doc('Endpoint da Simulacao')
    @simulacao_ns.expect(simulacao_request_model)
    @simulacao_ns.marshal_with(simulacao_response_model)
    def post(self):

        data = simulacao_request_parser.parse_args()

        valor = data["valorDesejado"]
        prazo = data["prazo"]

        if int(prazo) == 0:
            abort(404, error={'msg': 'Produto nao encontrado'})

        valor_maximo_tratado = func.coalesce(Produto.vr_maximo, 1000000000000)
        prazo_maximo_tratado = func.coalesce(Produto.nu_maximo_meses, 1000000000000)

        produto = Produto.query.filter(
            Produto.vr_minimo <= valor,
            valor_maximo_tratado >= valor,
            Produto.nu_minimo_meses <= prazo,
            prazo_maximo_tratado >= prazo,
        ).first()

        if produto:
            tb_price = tabela_price(valor, prazo, produto.pc_taxa_juros)
            tb_sac = tabela_sac(valor, prazo, produto.pc_taxa_juros)

            data_dict = {
                "codigoProduto": produto.co_produto,
                "descricaoProduto": produto.no_produto,
                "taxaJuros": float(produto.pc_taxa_juros),
                "resultadoSimulacao": [
                    {"tipo": "SAC", "parcelas": tb_sac},
                    {"tipo": "PRICE", "parcelas": tb_price},
                ],
            }

            if current_app.config["TESTING"] is False:
                enviar_json_ao_eventhub(data_dict)

            return data_dict, 200

        logger.error({'msg': 'Produto nao encontrado'})
        abort(404, error={'msg': 'Produto nao encontrado'})
